Scripts/Runtime/Entities/TaskDriver/TaskSet/TaskData/DataStream/Cancellation/CancelProgressFlowNode.cs

```diff
@@ -212,5 +212,89 @@ private bool CheckIfWillComplete(bool isStillProcessing, ref EntityProxyInstance
                 }
             }
         }
+
+        /// <summary>
+        /// A wrapper that provides access to data version(s) shared between <see cref="ITaskSetOwner"/> instances.
+        /// This allows multiple <see cref="CancelProgressFlowNode"/>s that point to the same <see cref="ITaskSetOwner"/>
+        /// instance to track whether data for a given version has been processed.
+        ///
+        /// Call <see cref="Dispose()"/> on the instance when it is no longer required to release shared version
+        /// references.
+        /// </summary>
+        /// <remarks>
+        /// Example:
+        /// Multiple <see cref="CancelProgressFlowNode"/>s point to each <see cref="TaskDriverSystem{TTaskDriverType}"/>
+        /// instance.
+        /// </remarks>
+        private struct OwnerSharedDataVersion : IDisposable
```

Comment:
I don't have a problem with this, but curious why this is a struct and not a class? 

It only gets used in managed land? Might be more simple and not need the private inner class, could use the instance as `readonly` in the `CancelProgressFlowNode` etc.

Reply:
struct because it's lighter weight and not necessary. Less burden on the GC.

I'll admit, this is a bit convoluted but I'm worried about a set of potential errors if the consumer of the shared/ref counted value calls dispose or release multiple times. Having the wrapper/handle sort of type allows us to protect against Dispose being called multiple times and breaking the ref count.

Rather than a ref count, the alternative would be to maintain a `HashSet` of instances that have acquired the shared value. Then on release/dispose we only decrement the count if the instance exists in the collection and remove it. This approach seemed heavier weight than the wrapper/handle  struct though.

Reply:
> struct because it's lighter weight and not necessary. Less burden on the GC.

I think it's the same weight no? A class would just be a pointer, this struct just wraps a reference to a class which would just be the pointer. 

I would agree less burden on the GC if we were created and destroying throughout the lifecycle, but these will get created once at startup and disposed once at teardown. 

> I'll admit, this is a bit convoluted but I'm worried about a set of potential errors if the consumer of the shared/ref counted value calls dispose or release multiple times. Having the wrapper/handle sort of type allows us to protect against Dispose being called multiple times and breaking the ref count.
>
> Rather than a ref count, the alternative would be to maintain a HashSet of instances that have acquired the shared value. Then on release/dispose we only decrement the count if the instance exists in the collection and remove it. This approach seemed heavier weight than the wrapper/handle struct though.

Ok I see the value in having the second time you call dispose do a no-op. With that though we will have an error in the `Release` call, I've made a separate comment about it below.

The HashSet approach would work but yeah, you've already done this so it's fine. I think i was more concerned about us leaning in toward structs because we're so ingrained with DOTS now and this isn't really the place where it makes a difference and it pushes us out of our patterns for OO land with the `readonly` etc. 

I think you could accomplish the same thing with this being a class and having the static inner dictionary for the record-keeping. `PersistentDataSystem` does this today. 

The issue with multiple dispose calls could be mitigated by just checking `IsDisposed` before decrementing. 

All that said, this is here and it works, so I'm cool with it staying like this.

Reply:
I thought about this some more and there isn't enough safety provided by the struct wrapper to justify using it over class. I think the GC pressure and number of indirect lookups is still a mild concern but I shouldn't be pre-optimising.

`OwnerSharedDataVersion` is now a class.